runtime/runtime/src/congestion_control.rs

```diff
@@ -368,7 +376,63 @@ pub fn bootstrap_congestion_info(
     }))
 }
 
-fn receipt_size(receipt: &Receipt) -> Result<usize, IntegerOverflowError> {
+impl DelayedReceiptQueueWrapper {
+    pub fn new(queue: DelayedReceiptQueue) -> Self {
+        Self {
+            queue,
+            new_delayed_gas: 0,
+            new_delayed_bytes: 0,
+            removed_delayed_gas: 0,
+            removed_delayed_bytes: 0,
+        }
+    }
+
+    pub(crate) fn push(
+        &mut self,
+        trie_update: &mut TrieUpdate,
+        receipt: &Receipt,
+        config: &RuntimeConfig,
+    ) -> Result<(), RuntimeError> {
+        let delayed_gas = receipt_congestion_gas(receipt, &config)?;
+        let delayed_bytes = receipt_size(receipt)? as u64;
+        self.new_delayed_gas = safe_add_gas(self.new_delayed_gas, delayed_gas)?;
+        self.new_delayed_bytes = safe_add_gas(self.new_delayed_bytes, delayed_bytes)?;
+        self.queue.push(trie_update, receipt)?;
+        Ok(())
+    }
+
+    pub(crate) fn pop(
+        &mut self,
+        trie_update: &mut TrieUpdate,
+        config: &RuntimeConfig,
+    ) -> Result<Option<Receipt>, RuntimeError> {
+        let receipt = self.queue.pop(trie_update)?;
+        if let Some(receipt) = &receipt {
+            let delayed_gas = receipt_congestion_gas(receipt, &config)?;
+            let delayed_bytes = receipt_size(receipt)? as u64;
+            self.removed_delayed_gas = safe_add_gas(self.removed_delayed_gas, delayed_gas)?;
+            self.removed_delayed_bytes = safe_add_gas(self.removed_delayed_bytes, delayed_bytes)?;
+        }
+        Ok(receipt)
+    }
+
+    pub(crate) fn len(&self) -> u64 {
+        self.queue.len()
+    }
+
+    pub(crate) fn apply_congestion_changes(
+        self,
+        congestion: &mut CongestionInfo,
+    ) -> Result<(), RuntimeError> {
+        congestion.add_delayed_receipt_gas(self.new_delayed_gas)?;
+        congestion.remove_delayed_receipt_gas(self.removed_delayed_gas)?;
+        congestion.add_receipt_bytes(self.new_delayed_bytes)?;
+        congestion.remove_receipt_bytes(self.removed_delayed_bytes)?;
```

Comment:
Out of curiosity why do you keep `added` and `removed` separately? Can this just be a single `diff` field? No big deal by any means, this is totally fine too.

Reply:
It's pretty much just to avoid integer conversions. Integer arithmetic is just always a security nightmare, as seen by the gas accounting error we someone reported last year in the runtime and had been there for more than 2 years. That's why I go out of my way to avoid it, basically. :)

The obvious alternatives would be to use a `i64` that can go negative. It would decrease the limit for overflows but I don't think it would matter in practice. However, the code would have to convert from i64 to u64 and the other way around in a few additional places, which means more panic conditions in the code where we rely on implicit assumptions that we never have more than X gas in a single chunk.

I prefer using two signed integers which can keep the number of integer arithmetic that depends on external inputs to a minimum.

Reply:
I also added a quick note about that in the struct comment